### ISSUE TITLE
AUC and Loss discrepancy between transformer_moe.py and transformer.py

### ISSUE THREAD

**Issue by etragas-fathom** [tensorflow/tensor2tensor#1554]
## Background
We're attempting to reproduce the behaviour of transformer.py by using transformer_moe.py with the 'a' layer type.

## Experimental Setup
We launched two training runs, which differed only in model type and hparams. 

* Transformer run:
    * Model: [TransformerEncoder](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/transformer.py#L1232)
    * Hparams: [transformer_tiny](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/transformer.py#L1997)
* Transformer_moe.py run:
    * Model: [TransformerMoe](https://github.com/tensorflow/tensor2tensor/blob/f9f63c3f3c5aab8ed8070f10662ad8012c09c17b/tensor2tensor/models/research/transformer_moe.py#L56)
    * Hparams: We took [transformer_moe_base](https://github.com/tensorflow/tensor2tensor/blob/f9f63c3f3c5aab8ed8070f10662ad8012c09c17b/tensor2tensor/models/research/transformer_moe.py#L267) and then added [transformer_tiny's](https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/models/transformer.py#L1997) hparams on top, overwriting any values that were already defined `transformer_moe_base`. We also set the layer type to `a/a#`
    
## Expected behaviour

By setting transformer_moe's layer type to `a`, we expected to reproduce the behaviour of transformer.py

## Results
The `transformer_moe` run achieves its best performance at 250k steps with a loss of 0.00119; we let it run another 150k steps, but the loss never improves, hovering between 0.0012 and 0.0013. Our `transformer.py` run had a loss of 0.0012 at 250k steps, but unlike `transformer_moe.py` continues to improve. By step 400k its loss is 0.0010 and it goes on to train until step 1.5m achieving a loss of 0.0007.

## Thoughts
We suspect that the difference in performance is due to a difference between the behaviour of the `a` layer in transformer_moe.py and the multihead attention function in transformer.py

It looks like transformer_moe's 'a' layer type turns into multihead_attention as per https://github.com/tensorflow/tensor2tensor/blob/f9f63c3f3c5aab8ed8070f10662ad8012c09c17b/tensor2tensor/layers/common_attention.py#L280 which is ostensibly the same attention fn that used in transformer.py https://github.com/tensorflow/tensor2tensor/blob/master/tensor2tensor/layers/transformer_layers.py#L186

## Questions
Is the discrepancy in performance expected when using the `a` layer? If not, is transformer_moe.py's implementation of `a` missing some critical arguments that transformer.py passes that moe doesn't?
